src/gadgets/ecc.rs

```diff
@@ -160,6 +160,74 @@ impl<C: CurveAffine<Base = F>, F: PrimeFieldBits, const T: usize> EccChip<C, F,
         Ok(acc)
     }
 
+    // optimization version for scalar_mul
+    // we assume the point p0 is not infinity here
+    pub fn scalar_mul_opt(
```

Comment:
Please, explain the logic of the additional method. Why is it not an optimization of the original method? And if they are different in logic, the difference in name should not be with `_opt` prefix

Reply:
I mean explain in rustdoc, not here 😅

Reply:
It is the optimization of the original method. But the difference is that it is only correct when p0 is not infinity. The original `scalar_mul` method is correct in all cases.

Reply:
In that case: `scalar_mul_no_inf`. If you don't like this title we we can add wrapper like 'NotInfAssignedPoint` (like `NonZeroUsize` from str lib) and make overload for `AssignedPoint` & `NoInfAssignedPoint` via trait with only one method `scalar_mul`

If you prefer last option, let me know I will help you with impl

Reply:
We might have different limitations when we do optimizations like this, so we can keep the "_opt" naming.  As for adding the trait and NoInfAssignedPoint, let me think if it is necessary since we already have the check and will return Error.

Reply:
Meaning that `_opt`, doesn't tell me anything at all. It's an empty name. When should I use it? When should I not use it? Every time a programmer has to go read rustdoc to figure out if he should use a method, it's a loss, we should give maximum information by name or data types, because autocomplete gives them.

Reply:
There are two types of optimization, one is universal without restriction, we can just put it into scalar_mul, one is with some precondition like this one (non-zero). We can add any optimization of second type in this function and use a single name "_opt" to it.

Reply:
Another option is to add `_unsafe` and comments to describe the conditions

Reply:
We need to convey a bit of information to the developer and it's not that the method is optimized, it's when we can use it. A developer typing `scalar_mul` should get two options in the IDE window and by the name understand what options he has. If he sees `_opt` there he should select this option "to try", go to the method signature and read what exactly `_opt` is there. If he makes a mistake, doesn't read it, or reads it inattentively, there will be an error. 

That's why the name explicitly states the most important part, and here it is that the set of input invariants is bounded, not that the method is optimized.

Let me then write a wrapper with `NonInfAssignedPoint` and do a merge into this PR to avoid common names. Then there will only be `scalar_mul` with different types for the first operand

Reply:
It's okay to add a wrapper for this specific one, but later the complexity will probably grow. To make things simpler, what about add "_unsafe" and make it private to use just on the folding_E part.

Reply:
`unsafe` is also a controversial postfix, as it will confuse any developer, for in rust `unsafe` means something completely different. Moreover, this method is quite safe and normal, just it should be used when the point is guaranteed not to be 0.

I.e. it's not some complicated existential issue, all we need to do is map that bit of information outward in an obvious way

Reply:
I agree with Mikhail. Naming should tell when to use it, e.g. '_non_zero'; '_opt' only indicates it's faster and could lead to error if used under the wrong conditions.

Reply:
If we name it as _non_zero, what to name it when we add other optimization later?  We can make it private function as internal use (specific for folding).

Reply:
If we do optimizations that don't involve a non-zero point, we will have to add them to both functions and so renaming won't be as important.

If there will be some other set of additional conditions, we will discuss them in detail

Optimization is an internal matter, while constraint on input data is external.

Reply:
And the degree of publicity of the function does not affect the name. As long as we have no consumers other than ourselves.

Reply:
If we have new conditions, backwards compatibility rules require that they be added separately, not in the same function. 

But we don't need backward compatibility, so we can remove the old method and add a completely new one, trying to express the new set of conditions in the form of name and data parameters, solving a more complex problem.

Even if there are 10 conditions, choosing a generic _opt term is still a bad solution, the design would have to be complicated to express all the constraints on that data.

Reply:
Agree with @cyphersnake on 1) "backwards compatibility rules require that they be added separately" 2) "express the new set of conditions in the form of name and data parameters" 3) "Even if there are 10 conditions, choosing a generic _opt term is still a bad solution"

If we encounter new conditions in the future, we figure out the right solution for the new conditions at a later time; instead of using a generic but vague name preparing to add any stuff into it.

Reply:
change to "_non_zero"